tests/08-vxlan/01-vxlan.clab.yml

```diff
@@ -0,0 +1,17 @@
+name: vxlan
+
+topology:
+  nodes:
+    s1:
+      kind: srl
+      image: ghcr.io/nokia/srlinux
+      startup-config: 01-vxlan-s1.config
+  links:
+    - type: vxlan
+      endpoint:
+        node: s1
+        interface: e1-1
+        mac: 02:00:00:00:00:04
+      remote: 192.168.66.1
+      vni: 100
+      udp-port: 5555
```

Comment:
I think instead of this lab we can use the following:

```yaml
name: vxlan-embed

topology:
  nodes:
    l1:
      kind: linux
      image: alpine:3
      exec:
        - ip addr add dev eth1 192.168.0.1/30
      mgmt-ipv4: 172.20.20.91
    l2:
      kind: linux
      image: alpine:3
      exec:
        - ip addr add dev eth1 192.168.0.2/30
      mgmt-ipv4: 172.20.20.92

  links:
    - type: vxlan
      endpoint:
        node: l1
        interface: l1eth1
      remote: 172.20.20.92
      vni: 100
    - type: vxlan
      endpoint:
        node: l2
        interface: l2eth1
      remote: 172.20.20.91
      vni: 100
```

That way you don't have to touch the host at all (and clean it up, which is not happening currently)

Reply:
looking at the code it seems this lab is not possible to deploy. Since we have to find the outgoing interface (I wonder if this is a mandatory step?) we rely on the host's routing stack. It makes sense, since in real life applications, the tunnels are established between host nodes, not from within the containers, as in the lab above.

Reply:
It is not possible to deploy this one, since vxlan links created in a single netns can't have the same VNI.
I will rework this lab

Reply:
This lab works, but with prereq.
When the vlxan interface is first created in the host netns and then moved to the container netns, it needs to initiate connection towards the remote first.
If the remote end (`l2`) starts to ping `l1` first, it won't succeed. l1 needs to start pinging first.

```yaml
name: vxlan-embed

topology:
  nodes:
    l1:
      kind: linux
      image: alpine:3
      exec:
        - ip addr add dev eth1 192.168.0.1/30
      mgmt-ipv4: 172.20.20.91
    l2:
      kind: linux
      image: alpine:3
      exec:
        - >
          ash -c '
          apk add iproute2 &&
          ip link add name vxlan0 type vxlan id 100 remote 172.20.20.91 dstport 14789 &&
          ip l set dev vxlan0 up &&
          ip addr add dev vxlan0 192.168.0.2/30'
      mgmt-ipv4: 172.20.20.92

  links:
    - type: vxlan
      endpoint:
        node: l1
        interface: eth1
      remote: 172.20.20.92
      vni: 100
```

Reply:
There is something strange happening with vxlan tunnels moved to container netns.

I have built a topology that it's deployed in two different hosts like this:

```yaml title="Host1"

#host1
name: vxlan-embed
prefix: ""

mgmt:
  network: fixedips
  ipv4_subnet: 100.103.1.0/24

topology:
  kinds:
    linux:
      image: ghcr.io/hellt/network-multitool

  nodes:
    node1:
      kind: linux
      mgmt_ipv4: 100.103.1.11
      exec:
        - ip addr add dev eth1 192.168.0.1/30
  
  links:
    - type: vxlan
      endpoint:
        node: node1
        interface: eth1
        mac: 02:00:00:00:00:02
      remote: 100.103.2.22
      vni: 100
      udp-port: 4789
```

```yaml title="Host1"
#host2
name: vxlan-embed
prefix: ""

mgmt:
  network: fixedips
  ipv4_subnet: 100.103.2.0/24

topology:
  kinds:
    linux:
      image: ghcr.io/hellt/network-multitool

  nodes:
    node2:
      kind: linux
      mgmt_ipv4: 100.103.2.22
      exec:
        - ip addr add dev eth1 192.168.0.2/30
  
  links:
    - type: vxlan
      endpoint:
        node: node2
        interface: eth1
        mac: 02:00:00:00:00:01
      remote: 100.103.1.11
      vni: 100
      udp-port: 4789
```

ping from 192.168.0.1 to 192.168.0.2 and vice versa does not work.

The first thing I see is that vxlan tunnels are created with the “nolearning” flag. This disables the snooping of incoming packets that builds MAC to VTEP address fdb table.

```bash
2829: eth1@if2829: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 02:00:00:00:00:01 brd ff:ff:ff:ff:ff:ff link-netnsid 0 promiscuity 0 minmtu 68 maxmtu 65535
    vxlan id 100 remote 100.103.6.11 dev if30 srcport 0 0 dstport 4789 nolearning ttl auto ageing 300 udpcsum noudp6zerocsumtx noudp6zerocsumrx numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535
    inet 192.168.0.1/30 scope global eth1
```

If we look at fdbs in node1 and node2:

```bash
host1:
00:00:00:00:00:00 dev eth1 dst 100.103.2.22 via ifindex 9 link-netnsid 0 self permanent

host2:
00:00:00:00:00:00 dev eth1 dst 100.103.1.11 via ifindex 30 link-netnsid 0 self permanent
```

BUM traffic will be delivered to those dst IP addresses.
These ifindexes (Nexthops) are root namespace interfaces. They are the interfaces that are used to reach destination IP tunnel. It can be the mgmt. bridge or other interface in the host.

If we do a tcpdump to check vxlan traffic, we see that it’s not coming from “mgmt_ipv4” of the node but from the IP of the bridge in the host. It looks that the udp socket of the vxlan tunnel is created at the root namespace. I understand that’s the reason you also see “udp port 4789 unreachable” in the tcpdumps of container.

If I add fdb entries, using IPs in the root namepspace:

```bash

##host1
ip netns exec node1  bridge fdb append 00:00:00:00:00:00 dev eth1 self dst 100.103.2.1 vni 100 port 4789

##host2
ip netns exec node1 bridge fdb append 00:00:00:00:00:00 dev eth1 self dst 100.103.1.1 vni 100 port 4789
```

.. traffic starts to flow immediately.

Hope it helps.

Reply:
Hi @michelredondo 
We will bring back the learning flag back.
As for the interface IP - it is indeed expected that the interface belongs to the root netns, as this is where the tunnel interface was created.

Reply:
Even with the learning flag, topologies like the one I described will not work. Both containers have the  “00:00:00:00:00:00” fdb entry pointing to the incorrect IP, so it will never be possible to learn MACs from the other tunnel end. 

The https://github.com/srl-labs/containerlab/blob/main/tests/08-vxlan/01-vxlan.clab.yml test case works because srl1 fdb entry points to the correct IP (172.20.25.22). When you originate traffic from that side, srl1 sends BUM traffic to l2 node and l2 learns how to send traffic back.
If you try to originate traffic from l2 node, BUM traffic is sent to srl1 mgmt IP (172.20.25.21), but the udp vxlan socket is not listening there. It’s listening in root namespace (bridge IP: 172.20.25.1). Learning in this case never happens.

The only way I can think to make this work is to add an additional fdb entry pointing to the IP address in the root ns.  Maybe we could add this as an optional parameter in the link definition.